crates/sui-types/src/error.rs

```diff
@@ -417,3 +433,193 @@ impl ExecutionStateError for SuiError {
+    CircularObjectOwnership,
+    MissingObjectOwner,
+    InvalidSharedChildUse,
+
+    ModulePublishFailure,
+    ModuleVerificationFailure,
+
+    // Move Error
+    VmError,
+}
+
+impl std::fmt::Display for ExecutionErrorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ExecutionErrorKind: {:?}", self)
+    }
+}
+
+impl std::error::Error for ExecutionErrorKind {}
+
+#[derive(Debug)]
+pub struct ExecutionError {
+    inner: Box<ExecutionErrorInner>,
+}
+
+#[derive(Debug)]
+struct ExecutionErrorInner {
+    kind: ExecutionErrorKind,
+    source: Option<BoxError>,
+}
+
+impl ExecutionError {
+    pub fn new(kind: ExecutionErrorKind, source: Option<BoxError>) -> Self {
+        Self {
+            inner: Box::new(ExecutionErrorInner { kind, source }),
+        }
+    }
+
+    pub fn new_with_source<E: Into<BoxError>>(kind: ExecutionErrorKind, source: E) -> Self {
+        Self::new(kind, Some(source.into()))
+    }
+
+    pub fn from_kind(kind: ExecutionErrorKind) -> Self {
+        Self::new(kind, None)
+    }
+
+    pub fn kind(&self) -> &ExecutionErrorKind {
+        &self.inner.kind
+    }
+
+    pub fn to_execution_status(&self) -> ExecutionFailureStatus {
+        match self.kind() {
+            ExecutionErrorKind::InsufficientGas => ExecutionFailureStatus::InsufficientGas,
+            ExecutionErrorKind::TransferUnowned
+            | ExecutionErrorKind::TransferNonCoin
+            | ExecutionErrorKind::TransferInsufficientBalance
+            | ExecutionErrorKind::InvalidTransactionUpdate
+            | ExecutionErrorKind::ObjectNotFound
+            | ExecutionErrorKind::DeleteObjectOwnedObject
+            | ExecutionErrorKind::FunctionNotFound
+            | ExecutionErrorKind::ModuleNotFound
+            | ExecutionErrorKind::InvalidFunctionSignature
+            | ExecutionErrorKind::InvalidFunctionVisibility
+            | ExecutionErrorKind::InvalidNonEntryFunction
+            | ExecutionErrorKind::ExecutionInvariantViolation
+            | ExecutionErrorKind::TypeError
+            | ExecutionErrorKind::CircularObjectOwnership
+            | ExecutionErrorKind::MissingObjectOwner
+            | ExecutionErrorKind::InvalidSharedChildUse
+            | ExecutionErrorKind::ModulePublishFailure
+            | ExecutionErrorKind::ModuleVerificationFailure => {
+                ExecutionFailureStatus::MiscellaneousError
+            }
+            ExecutionErrorKind::VmError => {
+                let source = if let Some(source) = self.source() {
+                    source
+                } else {
+                    return ExecutionFailureStatus::MiscellaneousError;
+                };
+
+                if let Some(vmerror) = source.downcast_ref::<VMError>() {
+                    match (
+                        vmerror.major_status(),
+                        vmerror.sub_status(),
+                        vmerror.location(),
+                    ) {
```

Comment:
Should we use VMStatus here? with the keep/discard mechanism?

Not sure how we feel about people paying for bugs or not. I think forcing people to pay for bugs is fine enough, if we are consistent about it

Reply:
For clarity, referring to this one:
https://github.com/move-language/move/blob/main/language/move-core/types/src/vm_status.rs#L154

Reply:
I mean this bit of code is inspired by vm status but idk if we want to use it directly

Reply:
Today we don't have a notion of not paying if a txn makes it this far and idk if i want to open up that discussion yet :)